core/lib/zksync_core/src/commitment_generator/mod.rs

```diff
@@ -0,0 +1,264 @@
+use std::time::Duration;
+
+use anyhow::Context;
+use itertools::Itertools;
+use metrics::{CommitmentStage, METRICS};
+use multivm::zk_evm_latest::ethereum_types::U256;
+use tokio::{sync::watch, task::JoinHandle};
+use zksync_commitment_utils::{bootloader_initial_content_commitment, events_queue_commitment};
+use zksync_dal::ConnectionPool;
+use zksync_types::{
+    commitment::{AuxCommitments, CommitmentCommonInput, CommitmentInput, L1BatchCommitment},
+    writes::{InitialStorageWrite, RepeatedStorageWrite, StateDiffRecord},
+    L1BatchNumber, ProtocolVersionId, StorageKey, H256,
+};
+use zksync_utils::h256_to_u256;
+
+mod metrics;
+
+const SLEEP_INTERVAL: Duration = Duration::from_millis(100);
+
+#[derive(Debug)]
+pub struct CommitmentGenerator {
+    connection_pool: ConnectionPool,
+    stop_receiver: watch::Receiver<bool>,
+}
+
+impl CommitmentGenerator {
+    pub fn new(connection_pool: ConnectionPool, stop_receiver: watch::Receiver<bool>) -> Self {
+        Self {
+            connection_pool,
+            stop_receiver,
+        }
+    }
+
+    async fn calculate_aux_commitments(
+        &self,
+        l1_batch_number: L1BatchNumber,
+        protocol_version: ProtocolVersionId,
+    ) -> anyhow::Result<AuxCommitments> {
+        let mut connection = self
+            .connection_pool
+            .access_storage_tagged("commitment_generator")
+            .await?;
+        let events_queue = connection
+            .blocks_dal()
+            .get_events_queue(l1_batch_number)
+            .await?
+            .context("Events queue is required for post-boojum batch")?;
+        let initial_bootloader_contents = connection
+            .blocks_dal()
+            .get_initial_bootloader_heap(l1_batch_number)
+            .await?
+            .context("Bootloader initial heap is missing")?;
```

Comment:
You can drop `connection` after here so that it isn't held during computationally expensive tasks.